Add support for building with clang-cl

solemnwarning opened the pull request:
Both x86_32 and x86_64 build using the appropriate toolchain from VS2022, all tests pass except for the following in both cases:

```
System Certificate Store - Find Certificate by UTF8 subject DN ran 1 tests in 1.04 msec 1 FAILED
Failure 1: found certificate was nullopt (at src/tests/test_certstor_system.cpp:306)
```

solemnwarning commented:
It turns out the aforementioned test failure isn't specific to the clang-cl build - it happens with MSVC too, I think its failing because my Windows VM doesn't have the "D-TRUST Root Class 3 CA 2 EV 2009" certificate the tests are looking for.

reneme commented:
That's certainly possible. If I recall correctly, the certstore tests are looking for more than one cert. so if it fails for just that one, you're probably right.

reneme reviewed (commented):
Regarding the failing test: Let's try to assume the existence of a different certificate. Apparently the D-Trust certificate isn't as wide-spread.

See: https://github.com/randombit/botan/pull/4280

reneme commented:
@solemnwarning could you rebase to master and see if the test succeeds on your platform now?

solemnwarning commented:
@reneme

I rebased onto master and the tests still failed... they do pass if I import the "D-TRUST Root Class 3 CA 2 EV 2009" certificate into the trust store though, so clearly my test VM is lacking certificates that everyone else apparently has.

Regarding `/Zc:preprocessor`, it still builds okay, so I assume its fine without?

reneme commented:
> Regarding `/Zc:preprocessor`, it still builds okay, so I assume its fine without?

I added that for a change in the tests. So if the unit tests (`make tests`) still build fine, I'd also conclude that it's not needed.

> they do pass if I import the "D-TRUST Root Class 3 CA 2 EV 2009"

:+1:

reneme commented on src/lib/math/bigint/big_ops2.cpp:
```diff
@@ -11,6 +11,20 @@
+ * configure.py to allow cpu-specific libs in the compiler info or os+cc+arch-specific libs in the
+ * module info. Hopefully this can go away when the above Clang issue is fixed anyway.
+*/
+#if defined(_MSC_VER) && defined(_WIN64) && defined(__clang__)
+   #pragma comment(lib, "clang_rt.builtins-x86_64.lib")
+#endif
```
Unfortunately, this doesn't seem to work on my windows environment:

```
lld-link: error: undefined symbol: __udivti3
>>> referenced by botan-3.lib(math_bigint_big_ops3.obj):(unsigned __int64 __cdecl Botan::operator%(class Botan::BigInt const &, unsigned __int64))
>>> referenced by botan-3.lib(math_bigint_big_ops2.obj):(public: unsigned __int64 __cdecl Botan::BigInt::operator%=(unsigned __int64))
>>> referenced by botan-3.lib(math_bigint_divide.obj):(void __cdecl Botan::vartime_divide(class Botan::BigInt const &, class Botan::BigInt const &, class Botan::BigInt &, class Botan::BigInt &))
ninja: build stopped: subcommand failed.
```

Configuring/Building like:

```
 ./configure --without-documentation --cc=clangcl --disable-shared-library --build-tool=ninja --compiler-cache=sccache --build-targets=static,tests
ninja tests
```

solemnwarning replied:
@reneme

How are you running the configure script? I'm using `python configure.py ...` and then `ninja` from the "x64 Native Tools Command Prompt for VS2022", I checked out your branch and that builds fine for me too.

Here's my compiler version, for comparison:

![image](https://github.com/user-attachments/assets/4e32187d-4c3a-41b9-a60b-8176e76011f5)

solemnwarning replied:
It builds in CI now, although its failing on something to do with installing afterwards.

reneme replied:
The patch below should fix the installation:

```patch
diff --git a/src/scripts/ci_check_install.py b/src/scripts/ci_check_install.py
index 5b4ec18f1..f6a695115 100755
--- a/src/scripts/ci_check_install.py
+++ b/src/scripts/ci_check_install.py
@@ -27,7 +27,7 @@ def verify_library(build_config):

     major_version = int(build_config["version_major"])

-    if build_config['compiler'] == 'msvc':
+    if build_config['compiler'] in ['msvc', 'clangcl']:
         expected_lib_format = r'^botan-%d\.(dll|lib)$' % (major_version)
     elif build_config['os'] == 'macos':
         expected_lib_format = r'^libbotan-%d\.(a|dylib)$' % (major_version)
```

reneme replied:
For reasons unknown my clang_rt doesn't seem to provide this builtin.

```
MEUSEL@BS001125 C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.38.33130\lib\x64
(0) > dumpbin /symbols .\clang_rt.builtins-x86_64.lib | Select-String "udiv"

00D 00000000 SECT4  notype ()    External     | __udivsi3
00D 00000000 UNDEF  notype ()    External     | __udivsi3
00E 00000000 SECT4  notype ()    External     | __udivmodsi4
012 00000000 SECT6  notype       Static       | $unwind$__udivmodsi4
015 00000000 SECT7  notype       Static       | $pdata$__udivmodsi4
00D 00000000 SECT4  notype ()    External     | __udivmoddi4
011 00000000 SECT6  notype       Static       | $unwind$__udivmoddi4
014 00000000 SECT7  notype       Static       | $pdata$__udivmoddi4
00D 00000000 SECT4  notype ()    External     | __udivdi3
00D 00000000 UNDEF  notype ()    External     | __udivmoddi4
00D 00000000 UNDEF  notype ()    External     | __udivmodsi4
00D 00000000 UNDEF  notype ()    External     | __udivmoddi4
00E 00000000 UNDEF  notype ()    External     | __udivmoddi4
```

reneme replied:
Never mind. Updating Visual Studio helped.

reneme commented on src/build-data/cc/clangcl.txt:
```diff
@@ -0,0 +1,102 @@
+macro_name CLANG
```
Probably we should make this `CLANGCL` to be able to distinguish vanilla clang from clang-cl. Much like that was [done for Xcode](https://github.com/randombit/botan/pull/4010) a few months ago.

Note that this change is incorporated in the patch I proposed on top.

reneme reviewed (commented):
I tested on my windows environment and unfortunately the linker workaround doesn't seem to work for me. Or am I doing something wrong?

Also: I added a few commits that, I think, are worthwhile to add here: https://github.com/reneme/botan/commits/clang-cl/

* a CI job on Windows, and
* dedicated detection of 'clang-cl' and its version.

reneme commented on src/lib/math/bigint/big_ops2.cpp:
```diff
@@ -11,6 +11,20 @@
+ * configure.py to allow cpu-specific libs in the compiler info or os+cc+arch-specific libs in the
+ * module info. Hopefully this can go away when the above Clang issue is fixed anyway.
+*/
+#if defined(_WIN64) && defined(BOTAN_BUILD_COMPILER_IS_CLANGCL)
+   #pragma comment(lib, "clang_rt.builtins-x86_64.lib")
+#endif
```
Is there a 32bit version of this workaround? Not that I think there's a lot of practical use for that in 2024, but perhaps we should try to add a 32bit clangcl build?

reneme replied:
Apparently that is relevant on 64 bit platforms only.  As per: https://stackoverflow.com/questions/68676184/clang-cl-error-lld-link-error-undefined-symbol-divti3#comment121371051_68676270

> GCC and Clang only support 128-bit integers in 64-bit mode, even in Linux. It's not because of the calling convention because large structs can be still be returned via a hidden pointer

randombit replied:
Might still be relevant for aarch64 though?

randombit replied:
Since the issue was opened in 2015 and is still unresolved I'm not holding out a huge amount of hope that it will ever be fixed. So I'd be inclined to try to address this within the build system. I can look at this part.

reneme commented on src/lib/tls/tls13_pqc/hybrid_public_key.h:
```diff
@@ -93,6 +93,16 @@ class BOTAN_TEST_API Hybrid_KEM_PrivateKey final : public Private_Key,
    public:
       Hybrid_KEM_PrivateKey(std::vector<std::unique_ptr<Private_Key>> private_keys);
 
+      /* Work around a bug/oddity in clang-cl on Windows - the default move assignment operator
+       * generates multiple calls to Public_Key(Public_Key&&) as a technically-valid-but-dangerous
+       * side-effect of the class having the dllexport attribute. In the cast of the Public_Key
+       * class, this isn't actually a problem since its just an interface and has no data members,
+       * but Clang still raises a warning for it.
+       *
+       * Hybrid_KEM_PrivateKey has an explicit move assignment operator to silence the warning.
+      */
+      Hybrid_KEM_PrivateKey &operator=(Hybrid_KEM_PrivateKey&& other);
```
Perhaps adhere to the rule-of-five then, and also add a copy c'tor, copy assignment and destructor that should all be 'default'. It seems that adding a non-default move c'tor would also make sense give the described oddity, no?

solemnwarning replied:
I've added explicit default/delete for the other move/copy assignment operators and c'tors for completeness, not inclined to add a custom move c'tor since its just a workaround for a compiler warning and not an actual issue (the warning doesn't get emited for the default move c'tors).

GCC is supposedly a bit more pragmatic here and won't emit the warning for the multiply-moved-trivially-movable-base-class case, so Clang might follow suit in the future...

reneme commented on src/lib/tls/tls13_pqc/hybrid_public_key.cpp:
```diff
@@ -349,6 +349,12 @@ Hybrid_KEM_PrivateKey::Hybrid_KEM_PrivateKey(std::vector<std::unique_ptr<Private
 }
 
+Hybrid_KEM_PrivateKey& Hybrid_KEM_PrivateKey::operator=(Hybrid_KEM_PrivateKey&& other) {
+   static_cast<Hybrid_KEM_PublicKey&>(*this) = static_cast<Hybrid_KEM_PublicKey&&>(other);
+   m_private_keys = std::move(other.m_private_keys);
+   return *this;
```
```suggestion
   if (this != &other) {
      static_cast<Hybrid_KEM_PublicKey&>(*this) = static_cast<Hybrid_KEM_PublicKey&&>(other);
      m_private_keys = std::move(other.m_private_keys);
   }
   return *this;
```

reneme commented on src/lib/tls/tls13_pqc/hybrid_public_key.cpp:
```diff
@@ -349,6 +349,12 @@ Hybrid_KEM_PrivateKey::Hybrid_KEM_PrivateKey(std::vector<std::unique_ptr<Private
 }
 
+Hybrid_KEM_PrivateKey& Hybrid_KEM_PrivateKey::operator=(Hybrid_KEM_PrivateKey&& other) {
+   static_cast<Hybrid_KEM_PublicKey&>(*this) = static_cast<Hybrid_KEM_PublicKey&&>(other);
```
That looks somewhat unusual. Could paste the warning clangcl did generate for the original code please?

solemnwarning replied:
```
build/include/public\botan/dh.h(80,30): error: defaulted move assignment operator of 'DH_PrivateKey' will move assign virtual base class 'Private_Key' multiple times [-Werror,-Wmultiple-move-vbase]
   80 | class BOTAN_PUBLIC_API(2, 0) DH_PrivateKey final : public DH_PublicKey,
      |                              ^
build/include/public\botan/dh.h(80,7): note: due to 'DH_PrivateKey' being dllexported
   80 | class BOTAN_PUBLIC_API(2, 0) DH_PrivateKey final : public DH_PublicKey,
      |       ^
build/include/public\botan/compiler.h(31,36): note: expanded from macro 'BOTAN_PUBLIC_API'
   31 | #define BOTAN_PUBLIC_API(maj, min) BOTAN_DLL
      |                                    ^
<command line>(3,30): note: expanded from macro 'BOTAN_DLL'
    3 | #define BOTAN_DLL __declspec(dllexport)
      |                              ^
build/include/public\botan/dh.h(81,52): note: 'Private_Key' is a virtual base class of base class 'PK_Key_Agreement_Key' declared here
   81 |                                                    public PK_Key_Agreement_Key,
      |                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
build/include/public\botan/dh.h(82,52): note: virtual base class 'Private_Key' declared here
   82 |                                                    public virtual Private_Key {
      |                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

reneme replied:
I think, instead of defining custom move constructors as a workaround, we actually want to add `virtual` inheritance to a few more base class declarations. See this patch: https://github.com/reneme/botan/commit/f663994ad84c62127d419b26c1787e5e9e016805

solemnwarning replied:
Much better!

solemnwarning commented on src/lib/prov/pkcs11/p11.h:
```diff
@@ -24,7 +24,14 @@
 #endif
 
-#if defined(_MSC_VER)
+/* When building with clang-cl, specifying this with dllimport causes a warning when defining the
+ * CK_C_XXX function pointer types in pkcs11.h:
+ *
+ * 'dllimport' attribute only applies to functions, variables, classes, and Objective-C interfaces
+ *
+ * I'm not convinced the dllimport is needed/approproate for typedefs under MSVC either.
+*/
+#if 0  // defined(_MSC_VER)
    #define CK_DECLARE_FUNCTION_POINTER(returnType, name) returnType __declspec(dllimport)(*name)
 #else
    #define CK_DECLARE_FUNCTION_POINTER(returnType, name) returnType(*name)
```
I'm not too sure either, but I am assuming any breakage will be really obvious linker or loader failures rather than anything subtle.

reneme replied:
Fair. And it does sound somewhat odd to "dllimport" a function pointer declaration (assuming it is not a global variable or static member variable that is defined in some dllexporting compilation unit). For the sake of caution, we could also special-case this for clang-cl, I guess:

```suggestion
// When building with clang-cl on Windows, specifying this with dllimport causes
// a warning when defining the CK_C_XXX function pointer types in pkcs11.h:
//
//   'dllimport' attribute only applies to functions, variables, classes, and Objective-C interfaces
//
// ... hence, we enable this for MSVC only.
#if defined(BOTAN_BUILD_COMPILER_IS_MSVC)
   #define CK_DECLARE_FUNCTION_POINTER(returnType, name) returnType __declspec(dllimport)(*name)
#else
   #define CK_DECLARE_FUNCTION_POINTER(returnType, name) returnType(*name)
```

@randombit, do you have an opinion here? Also regarding semver, perhaps?

Otherwise: if we all agree with not dllimport-declaring this any longer, I'd say we should get rid of the `#if 0` and just 
 leave the bare `#define` from the else-branch.

randombit replied:
Even the latest (that I can find) header in https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/os/include/pkcs11-v3.0/pkcs11.h has similar commentary. That said I agree it doesn't seem to make sense to "dllimport" this, *and* it does seem like if it is incorrect it would probably fail in an obvious (eg link error) vs subtle way.

I don't see any SemVer implications to this either way.

I guess I'd be ok with either leaving it as MSVC specific or just removing it entirely and using the `#else` branch everywhere.

reneme commented on src/lib/prov/pkcs11/p11.h:
```diff
@@ -40,7 +47,7 @@
 #endif
 
-#include "pkcs11.h"
+#include <pkcs11.h>
```
I can imagine the quote-based include might have been on purpose here. We do ship the upstream pkcs11 headers in the same include directory. There might be a case for preferring those over some system-provided ones. @randombit?

randombit replied:
I do not recall the reasoning on this one. I guess if we end up picking up a system provided header that's ok as long as it's recent enough.

reneme reviewed (approved):
Looks good to me, apart from the two open discussions in `p11.h`.

randombit commented on src/build-data/ninja.in:
```diff
@@ -10,7 +10,7 @@ LANG_EXE_FLAGS = %{cc_lang_binary_linker_flags}
 WARN_FLAGS     = %{cc_warning_flags}
 
-LDFLAGS        = %{ldflags} %{cc_compile_flags}
+LDFLAGS        = %{ldflags}
```
This behavior (passing the compiler flags also to the linker) was just added in #4206 which I guess is kind of inelegant but we'd need to figure out some alternate solution for #4196

randombit commented on configure.py:
```diff
@@ -2313,11 +2313,21 @@ def test_exe_extra_ldflags():
-    if cc.basename == 'msvc' and variables['cxx_abi_flags'] != '':
-        # MSVC linker doesn't support/need the ABI options,
-        # just transfer them over to just the compiler invocations
+    # On MSVC, the "ABI flags" should be passed to the compiler only, on other platforms, the
```
These shenanigans seem to suggest that we might end up with something that builds in CI but not on a developers machine, which isn't helpful. Is the issue here that Clang really dislikes `/MD`/etc during the link step?

TriceHelix replied:
This is relevant to https://github.com/randombit/botan/issues/4424.

solemnwarning commented:
@randombit @reneme

Are either of you anticipating me making any changes on this PR? The remaining points seem to be a bit "maybe do something".

reneme commented on src/build-data/cc/clangcl.txt:
```diff
@@ -0,0 +1,102 @@
+sse2   -> "-msse2"
+ssse3  -> "-mssse3"
+sse41  -> "-msse4.1"
+sse42  -> "-msse4.2"
```
Note that there's [a pull request that removes explicit (but unused) support for SSE 4.2](https://github.com/randombit/botan/pull/4331). This will likely cause issues on master. Please rebase your patch onto master once the linked PR is merged and simply remove this line.

reneme commented:
> Are either of you anticipating me making any changes on this PR?

This is awaiting approval by @randombit

reneme commented on src/build-data/cc/clangcl.txt:
```diff
@@ -0,0 +1,102 @@
+# These can be overridden with --msvc-runtime option
+rt               -> "/MD"
+rt-debug         -> "/MDd"
+</mach_abi_linking>
```
With #4447 likely being merged first, this will need a rebase and an additional section:

```
<ct_value_barrier>
default -> asm
</ct_value_barrier>
```

randombit commented:
@solemnwarning Sorry about the long wait here. This needs a rebase and squash but otherwise is probably ok to land. I'm not tremendously happy about the clang rt thing but it seems a longstanding issue and unlikely to be fixed anytime soon so I guess we have to just live with it. Once this is merged we can add a CI build and then iterate from there.

solemnwarning commented:
@randombit I've merged the changes from master with mercifully few conflicts and made a couple of further changes:

- Updated `src/build-data/cc/clangcl.txt` per https://github.com/randombit/botan/pull/4255#discussion_r1860111585
- Removed `BOTAN_PUBLIC_API` from `Botan::IPAddressBlocks::Version` definition as per the earlier change to `p11.h`.

randombit commented:
@solemnwarning I missed your message until now. There is another merge conflict 😅 hopefully minor. I will re-review asap.

solemnwarning commented:
@randombit updated

randombit commented:
@solemnwarning Last time I promise :sweat_smile: but can you rebase against `master` again [*] and also squash the commits? Just to a single commit with a reasonable description is fine. Beyond this LGTM/ready to merge for 3.10.

[*] There is no explicit merge conflict but there were a lot of changes to `clang-tidy` configuration in the last couple of months and I think it's possible to cause failures on master if we merge something that passes with the old `clang-tidy` config but causes errors in the new config.
